src/charmed_kubeflow_chisme/testing/cos_integration.py

```diff
@@ -0,0 +1,226 @@
+
+
+def _get_alert_rules(data: str) -> Set[str]:
+    """Get all alert rules from string, e.g. file content or relation data.
+
+    Example of relations data of metrics-endpoint would be:
+
+    ```Python
+    'alert_rules': '{"groups": [{rules": [{"alert": "my-alert", ...
+    ```
+
+    Example of rule file with single alert rule:
+
+    ```yaml
+    alert: my-alert
+    expr: up < 1
+    for: 5m
+    ...
+    ```
+
+    Example of rule file with multiple alert rules:
+
+    ```yaml
+    groups:
+    - name: my-group
+      rules:
+      - alert: my-alert
+    ...
+    ```
+    """
+    alert_rules = yaml.safe_load(data)
+    if "groups" in alert_rules:
+        return {rule["alert"] for group in alert_rules["groups"] for rule in group["rules"]}
+
+    return {alert_rules["alert"]}
+
+
+def _get_metrics_endpoint(data: str) -> Set[str]:
+    """Get set of metrics endpoints from string.
+
+    This function is excpection data defined as string.
+
+    ```json
+    [
+      {
+        "metrics_path": "/metrics",
+        "static_configs": [
+          {
+            "targets": [
+              "*:5000",
+              "*:8000"
+            ]
+          }
+        ]
+      }
+    ]
+    ```
+    """
+    metrics_endpoints = set()
+    scrape_jobs = yaml.safe_load(data)
+    for job in scrape_jobs:
+        path = job["metrics_path"]
+        metrics_endpoints |= {
+            f"{target}{path}" for config in job["static_configs"] for target in config["targets"]
+        }
+
+    return metrics_endpoints
+
+
+async def _run_on_unit(unit: Unit, cmd: str) -> Action:
+    """Run command on unit."""
+    log.info("running cmd `%s` on unit %s", cmd, unit.name)
+    result = await unit.run(cmd, block=True)  # Note(rgildein): Using block to wait for results
+
+    assert (
+        result.results["return-code"] == 0
+    ), f"cmd `{cmd}` failed with error `{result.results.get('stderr')}`"
+    return result
+
+
+def get_alert_rules(path: Path = ALER_RULES_DIRECTORY) -> Set[str]:
+    """Get all alert rules from files."""
+    alert_rules = set()
+    for file_type in ["*.rule", "*.rules"]:
+        for file in path.glob(file_type):
+            alert_rules |= _get_alert_rules(file.read_text())
+
+    return alert_rules
+
+
+async def assert_alert_rules(app: Application, alert_rules: Set[str]) -> None:
+    """Assert function comparing alert rules from relation data and provided aler rules.
+
+    This function compare alert rules defined in METRICS_ENDPOINT_RELATION relation data bag and
+    provided aler rules. e.g. {"my-alert1", "my-alert2"}
+    """
+    relation_data = await _get_app_relation_data(app, METRICS_ENDPOINT_RELATION)
+    assert "alert_rules" in relation_data, "relation is missing alert_rules"
+
+    relation_alert_rules = _get_alert_rules(relation_data["alert_rules"])
+
+    assert relation_alert_rules == alert_rules
+
+
+async def assert_metrics_endpoint(app: Application, metrics_endpoints: Set[str]) -> None:
```

Comment:
I 'd expect that the user could pass the `app` and separately (pairs of) `metrics_port` and `metrics_path` instead of the metrics endpoint directly. I think it makes for a more clear UX than having to construct the metrics endpoint themselves. Then, the function could construct the endpoint as `{app.name}.{app.model.name}.svc:{metrics_port}{metrics_path}`. It adds though the convention that all metrics endpoints use this [format](https://github.com/canonical/argo-operators/blob/main/charms/argo-controller/src/charm.py#L66-L67). That being said, I don't think any CKF charm follows a different pattern, thus if we mention that in the docstring, we could be good. WDYT?

Reply:
I have no strong opinion here, like this, it was just simpler. If we go with metrics_port and metrics_path, you need to create and combination, if any charm use more than one e.g. [mlflow-operator](https://github.com/canonical/mlflow-operator/blob/753192d5471d880ac4b9de5c3e7e7d1a71e8f090/src/charm.py#L88-L103). And this can become more complex if you want to use two ports with two different paths, that would be too complex to implement. 
We can go with single port and path per assert, like this:
```python
await assert_metrics_endpoint(mlflow, metrics_port=5000, metrics_path="/metrics")
await assert_metrics_endpoint(mlflow, metrics_port=8000, metrics_path="/metrics")
```
So most charm implementation will be just with single assert, and user can easily do multiple if required. WDYT?

Reply:
Yes, I like it +1

Reply:
Ok, but what about a situation like this [endpoint](https://github.com/canonical/charmed-kubeflow-chisme/blob/659a4494460bd83450d4212aa838ccd0d0f650be/tests/integration/test_cos.py#L50-L51)? I think it's not normal situation, but still it can happen. WDYT?